Bug 1848100 - Implement add-on installation failed listener.

Amejia481 opened the pull request:
🔴  Needs to land first https://phabricator.services.mozilla.com/D186413

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
<!-- Do not add anything below this line -->

Used by GitHub Actions.

Amejia481 commented:
A couple of examples, removed the title from the dialog until we could get into an agreement with the UX/UI team.

![Screenshot_20230817-230958](https://github.com/mozilla-mobile/firefox-android/assets/773158/c0daac7c-d668-43a2-9eeb-acf5757fd910)
![Screenshot_20230817-215916](https://github.com/mozilla-mobile/firefox-android/assets/773158/7ac3d011-33ab-4081-90ca-8ffc29f56a1e)
![Screenshot_20230817-215855](https://github.com/mozilla-mobile/firefox-android/assets/773158/f9df6ab7-e171-46d4-97ed-e62a489247ab)

Amejia481 commented:
This patch could replace the functionality on https://github.com/mozilla-mobile/firefox-android/pull/3276/files

willdurand commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,6 +173,8 @@
+    <!-- Text shown when attempting to install an add-on and a network error happened. -->
+    <string name="mozac_feature_addons_failed_to_install_network_error">The add-on could not be downloaded because of a connection failure</string>
     <!-- Text shown when attempting to install a blocklisted add-on. %1$s is the add-on name. -->
     <string name="mozac_feature_addons_blocklisted">%1$s could not be installed because it has a high risk of causing stability or security problems</string>
```
I tried to install a blocklisted extension from the Fenix manager and I get this error message, which is great, except that the `extension` is `null` (the one coming from GV) so we don't have the add-on name here.

Amejia481 replied:
Thanks for testing! Could you add the steps to reproduce for me to try to replicate it?

willdurand replied:
STR:

1. checkout https://github.com/Rob--W/android-ext-qa locally
2. go into the checked out directory and run: `adb push ext-qa-data/ /data/local/tmp/ext-qa-data`
3. open Firefox Nightly (Desktop) and connect to Fenix via the debugging protocol
4. run the code snippet in the privileged context:

```
IOUtils.copy(
  "/data/local/tmp/ext-qa-data/mozilla_components_addon_collection_en_Extensions-for-Android.json",
  Services.env.get("HOME") + "/mozilla_components_addon_collection_en_Extensions-for-Android.json"
);
```

Amejia481 replied:
I think it should be working with the new changes, would you mind re-testing ? :) 
![image](https://github.com/mozilla-mobile/firefox-android/assets/773158/360f817b-63b4-4af9-bb4c-1fb7912a1b6a)

willdurand commented on android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtensionException.kt:
```diff
@@ -22,9 +28,50 @@ class GeckoWebExtensionException(throwable: Throwable) : WebExtensionException(t
+                    )
+
+                    ERROR_NETWORK_FAILURE -> WebExtensionInstallException.NetworkFailure(
+                        extensionName = throwable.extensionName,
```
I thought we couldn't get the extension name when there was a network failure?

willdurand commented on android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionDelegate.kt:
```diff
@@ -110,6 +110,18 @@ interface WebExtensionDelegate {
+     * Invoked whenever the installation of a [WebExtension] failed.
+     *
+     * @param extension the extension that failed to be installed, can be null in cases where
+     * the extension couldn't downloaded or the extension couldn't be parsed.
```
```suggestion
     * the extension couldn't be downloaded or the extension couldn't be parsed.
```

Amejia481 replied:
📝

willdurand commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,6 +173,18 @@
+    <!-- Text shown when attempting to install an add-on and the downloaded file is corrupted. -->
+    <string name="mozac_feature_addons_failed_to_install_corrupt_error">This add-on could not be installed because it appears to be corrupt</string>
+    <!-- Text shown when attempting to install an add-on and the downloaded hash file is the wrong one. -->
+    <string name="mozac_feature_addons_failed_to_install_hash_error">This add-on could not be installed because it does not match the expected</string>
```
```suggestion
    <string name="mozac_feature_addons_failed_to_install_hash_error">This add-on could not be installed because it does not match the expected.</string>
```

willdurand commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,6 +173,18 @@
+    <!-- Text shown when attempting to install an add-on and a network error happened. -->
+    <string name="mozac_feature_addons_failed_to_install_network_error">The add-on could not be downloaded because of a connection failure</string>
+    <!-- Text shown when attempting to install an add-on and the downloaded file is corrupted. -->
+    <string name="mozac_feature_addons_failed_to_install_corrupt_error">This add-on could not be installed because it appears to be corrupt</string>
```
```suggestion
    <string name="mozac_feature_addons_failed_to_install_corrupt_error">This add-on could not be installed because it appears to be corrupt.</string>
```

willdurand commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,6 +173,18 @@
     <!-- Text shown after failed to install an add-on. %1$s is the add-on name. -->
     <string name="mozac_feature_addons_failed_to_install">Failed to install %1$s</string>
+    <!-- Text shown when attempting to install an add-on and a network error happened. -->
+    <string name="mozac_feature_addons_failed_to_install_network_error">The add-on could not be downloaded because of a connection failure</string>
```
```suggestion
    <string name="mozac_feature_addons_failed_to_install_network_error">The add-on could not be downloaded because of a connection failure.</string>
```

willdurand commented on android-components/components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt:
```diff
@@ -234,7 +235,28 @@ object WebExtensionSupport {
+                        )
+                        return
+                    }
+                    if (!extension.isBuiltIn()) {
```
is this actually happening? like, could this method be called for a built-in extension?

Amejia481 replied:
I'm not completely sure, just added the check just in case as it will be really odd for an users. Will try to simulated an installation error one of our built-in to see.

Amejia481 replied:
Verified with [dummy-incompatible.xpi](https://searchfox.org/mozilla-central/source/mobile/android/geckoview/src/androidTest/assets/web_extensions/dummy-incompatible.xpi) an xip we host on m-c. built-in extensions have a different error handling that other extension, we can remove the check as this code path is not executed :)

willdurand commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -51,26 +55,47 @@ class WebExtensionPromptFeature(
+            )
+        }
+    }
+    private fun handleBeforeInstallationRequest(promptRequest: WebExtensionPromptRequest.BeforeInstallation) {
+        when (promptRequest) {
+            is WebExtensionPromptRequest.BeforeInstallation.InstallationFailed -> {
+                handleInstallationFailedRequest(
+                    exception = promptRequest.exception,
+                )
+                consumePromptRequest()
+            }
+        }
+    }
```
```suggestion

        private fun handleBeforeInstallationRequest(promptRequest: WebExtensionPromptRequest.BeforeInstallation) {
        when (promptRequest) {
            is WebExtensionPromptRequest.BeforeInstallation.InstallationFailed -> {
                handleInstallationFailedRequest(
                    exception = promptRequest.exception,
                )
                consumePromptRequest()
            }
        }
    }
    
```

(blank lines)

Amejia481 replied:
Now I realized why I was missing so many blank lines, my IDE was giving me the wrong impression of an empty line with the git author information 😅 . 
 
![image](https://github.com/mozilla-mobile/firefox-android/assets/773158/421e825e-a6ed-4d70-8a81-dde11bc2f236)

Amejia481 replied:
Now it's accurate XD 
![image](https://github.com/mozilla-mobile/firefox-android/assets/773158/6f3320a6-bb90-4fda-9666-63c0c608824c)

willdurand commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -79,13 +104,76 @@ class WebExtensionPromptFeature(
+                    appName,
+                    version,
+                )
+            }
```
Shouldn't we have a default case?

Amejia481 replied:
Do you mean for the `when` expression, like an `else` statement ?

Amejia481 replied:
If so, it's not needed as we are using a [sealed class](https://kotlinlang.org/docs/sealed-classes.html#sealed-classes-and-when-expression) which in a [when expression](https://kotlinlang.org/docs/sealed-classes.html#sealed-classes-and-when-expression) enforce by the compiler that all cases are fulfilled.

willdurand reviewed (commented):
I only started a review but didn't finish yet, here is some preliminary feedback. Thanks!

Amejia481 commented:
I think I addressed all comments, I'll work on [bug #1851766](https://bugzilla.mozilla.org/show_bug.cgi?id=1851766) to see if we can remove the [hack](https://github.com/mozilla-mobile/firefox-android/pull/3332/files#diff-2ddc504c9aef157255a3ebc15ca5ccadbf966934d724e5a52f8db8b15da4180fR947) before landing.

Amejia481 commented:
As [bug #1851766](https://bugzilla.mozilla.org/show_bug.cgi?id=1851766) landed, I removed the [hack](https://github.com/mozilla-mobile/firefox-android/pull/3332/files#diff-2ddc504c9aef157255a3ebc15ca5ccadbf966934d724e5a52f8db8b15da4180fR947) :)

willdurand commented on android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:
```diff
@@ -930,6 +943,17 @@ class GeckoEngine(
         )
     }
+
+    internal fun org.mozilla.geckoview.WebExtension?.toSafeWebExtension(): GeckoWebExtension? {
```
Shouldn't we remove this?

Amejia481 replied:
We removed the [suppression](https://github.com/mozilla-mobile/firefox-android/commit/cbfc8d34e086769d2d8184ec311a1a1dec0ce16e#diff-2ddc504c9aef157255a3ebc15ca5ccadbf966934d724e5a52f8db8b15da4180fR949-R950) and the `this.id != null` check.

```kotlin
    @Suppress("SENSELESS_COMPARISON")
        return if (this != null && this.id != null) {
```

But still we need to verify that the gecko view extension is not null before try to create an AC extension :)

willdurand commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -79,13 +106,76 @@ class WebExtensionPromptFeature(
+                context.getString(R.string.mozac_feature_addons_failed_to_install_hash_error)
+            }
+
+            is WebExtensionInstallException.FileAccess -> {
```
I don't think this is possible on Android (it was mainly a Windows thing). Can we omit this error? We still have the `Unknown` case anyway.

Amejia481 replied:
Removed!

willdurand commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,6 +173,18 @@
+    <!-- Text shown when attempting to install an add-on and the downloaded hash file is the wrong one. -->
+    <string name="mozac_feature_addons_failed_to_install_hash_error">This add-on could not be installed because it does not match the expected.</string>
+    <!-- Text shown when attempting to install an add-on and an error occurred trying to write to the filesystem. %1$s is the add-on name. -->
+    <string name="mozac_feature_addons_failed_to_install_file_access_error">%1$s could not be installed because the needed file cannot be modified.</string>
```
See comment elsewhere about omitting this error entirely for now.

willdurand commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -79,13 +106,76 @@ class WebExtensionPromptFeature(
+                context.getString(R.string.mozac_feature_addons_failed_to_install_corrupt_error)
+            }
+
+            is WebExtensionInstallException.IncorrectHash -> {
+                context.getString(R.string.mozac_feature_addons_failed_to_install_hash_error)
+            }
```
I don't think we can easily run into this error since this seems to be only used for updates, which aren't triggered by users. Maybe we could omit this error. WDYT?

Amejia481 replied:
Removed :)

willdurand reviewed (changes requested):
Sorry, it took me a lot of time to review this PR but I verified every error message and wrote a document that we'll share with QA later. I have many comments so requesting changes for now.

Overall, everything works well, though so thank you!

Amejia481 commented on android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtensionException.kt:
```diff
@@ -22,9 +28,49 @@ class GeckoWebExtensionException(throwable: Throwable) : WebExtensionException(t
+                    )
+
+                    ERROR_SIGNEDSTATE_REQUIRED -> WebExtensionInstallException.UnSigned(
+                        extensionName = throwable.extensionName,
```
Should we just use omit the name and use **"This add-on could not be installed .... verified"**?

Amejia481 replied:
Added **"This add-on could not be installed .... verified"**.

Amejia481 commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -79,13 +106,76 @@ class WebExtensionPromptFeature(
+
+            is WebExtensionInstallException.Incompatible -> {
+                val appName = context.getString(R.string.app_name)
+                val version = context.appVersionName
```
I think that issue is because you are using a debug variant, same problem can be seen when going to the about page
<img width="393" alt="image" src="https://github.com/mozilla-mobile/firefox-android/assets/773158/5b818dd0-03fe-40df-8699-25804aca68c3">

When building locally a nightly version I get `119.0a1` for `context.appVersionName` :)

Amejia481 commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,6 +173,18 @@
+    <!-- Text shown when attempting to install an add-on and an error occurred because the extension was incompatible. %1$s is the add-on name, %2$s is the app name and %3$s is the app version. -->
+    <string name="mozac_feature_addons_failed_to_install_in_compatible_error">%1$s could not be installed because it is not compatible with %2$s %3$s.</string>
     <!-- Text shown when attempting to install a blocklisted add-on. %1$s is the add-on name. -->
     <string name="mozac_feature_addons_blocklisted">%1$s could not be installed because it has a high risk of causing stability or security problems</string>
```
Added!

Amejia481 commented:
Thanks for the review, I think I addressed all the comments. This is ready for a second pass!

willdurand commented on android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtensionException.kt:
```diff
@@ -22,9 +26,37 @@ class GeckoWebExtensionException(throwable: Throwable) : WebExtensionException(t
+                        throwable,
+                    )
+
+                    else -> WebExtensionInstallException.Unknown(
```
I am wondering if we should have the name of the add-on here. I know it _can_ be present but what if it isn't because of a random error that we don't explicitly support above? Wouldn't it be more robust to never consider the name of the extension in all _unknown_ cases?

Amejia481 replied:
>         val addonName = exception.extensionName ?: "" 


As we are falling back to an [empty string when the add-on name is empty](https://github.com/mozilla-mobile/firefox-android/pull/3332/files/bf56ba952c42b841beb2babd2eddc5993fc0a0c2#diff-c592bd7feeeae61a4cb385f42a7543d328e91f6d2823a05b3f430a8d4c0034f4R122), we are going to end up with the following strings:

1. **When the add-on name is not present** "_Failed to install "_.
2. **When it's present** _"Failed to install add-on name"_.

willdurand replied:
Ok, I guess that's fine then, thanks! (At least in English..)

Amejia481 replied:
> (At least in English..)

That's a good point, I'll remove the name just to be sure!

Amejia481 replied:
> Do you think we could add a second string that is only "Failed to install this add-on." when we do not have a extension name?

Yes, let's do that!

willdurand commented on android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt:
```diff
@@ -5,32 +5,56 @@
-     * Value type that represents a request for a permission prompt.
-     * @property extension The [WebExtension] that requested the dialog to be shown.
-     * @property onConfirm A callback indicating whether the permissions were granted or not.
+     * Value type that represents a request for showing a native dialog from a [WebExtension] before
+     * the installation succeed.
```
```suggestion
     * Value type that represents a request for showing a native dialog from a [WebExtension] before
     * the installation succeeds.
```

willdurand commented on android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt:
```diff
@@ -5,32 +5,56 @@
-     * Value type that represents a request for showing post-installation prompt.
-     * Normally used to give users an opportunity to enable the [extension] in private browsing mode.
-     * @property extension The installed extension.
+     * Value type that represents a request for showing a native dialog from a [WebExtension] after
+     * installation succeed.
+     *
```
```suggestion
     * Value type that represents a request for showing a native dialog from a [WebExtension] after
     * installation succeeds.
     *
```

willdurand commented on android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt:
```diff
@@ -484,16 +484,50 @@ open class WebExtensionException(throwable: Throwable, open val isRecoverable: B
+        WebExtensionInstallException(throwable = throwable, extensionName = null)
+
+    /**
+     * The extension install was cancelled because must be signed and isn't.
```
```suggestion
     * The extension install was cancelled because the file must be signed and isn't.
```

willdurand commented on android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt:
```diff
@@ -484,16 +484,50 @@ open class WebExtensionException(throwable: Throwable, open val isRecoverable: B
+        WebExtensionInstallException(throwable = throwable)
+
+    /**
+     *  The extension install failed because a network error.
```
```suggestion
     *  The extension install failed because of a network error.
```

willdurand commented on android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionDelegate.kt:
```diff
@@ -110,6 +110,18 @@ interface WebExtensionDelegate {
+    /**
+     * Invoked whenever the installation of a [WebExtension] failed.
+     *
+     * @param extension the extension that failed to be installed, can be null in cases where
+     * the extension couldn't be downloaded or the extension couldn't be parsed.
```
```suggestion
     * @param extension the extension that failed to be installed. It can be null when the
     * extension couldn't be downloaded or the extension couldn't be parsed for example.
```

willdurand commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -79,13 +106,64 @@ class WebExtensionPromptFeature(
+                context.getString(R.string.mozac_feature_addons_blocklisted, addonName)
+            }
+
+            is WebExtensionInstallException.Unknown, is WebExtensionInstallException.UserCancelled -> {
```
Why is `UserCancelled` getting an error message? I thought we didn't want any message when that happens?

Amejia481 replied:
It's simulating the same behaviour that we had when we had the snackbar. I guess we should remove it?

Amejia481 replied:
Removed!

Amejia481 commented:
@willdurand I updated the PR :)

willdurand commented on fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt:
```diff
@@ -79,13 +106,69 @@ class WebExtensionPromptFeature(
+            }
+
+            is WebExtensionInstallException.UserCancelled -> {
+                // We don't want to show an error message when users cancel installations.
```
```suggestion
                // We don't want to show an error message when users cancel installation.
```

willdurand commented on fenix/app/src/test/java/org/mozilla/fenix/extension/WebExtensionPromptFeatureTest.kt:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Test
+    fun `WHEN InstallationFailed is dispatched THEN handleInstallationFailedRequest`() {
```
```suggestion
    fun `WHEN InstallationFailed is dispatched THEN handleInstallationFailedRequest is called`() {
```

maybe? not sure how we should read these BDD-style test names.

willdurand commented on fenix/app/src/test/java/org/mozilla/fenix/extension/WebExtensionPromptFeatureTest.kt:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Test
+    fun `WHEN calling handleInstallationFailedRequest with CorruptedFile error THEN showDialog with the correct message`() {
```
```suggestion
    fun `WHEN calling handleInstallationFailedRequest with CorruptFile error THEN showDialog with the correct message`() {
```

willdurand reviewed (approved):
Looks good, thanks! Please add the new error message for unknown errors without an add-on name.

flodolo commented on android-components/components/feature/addons/src/main/res/values/strings.xml:
```diff
@@ -173,8 +173,18 @@
+    <string name="mozac_feature_addons_failed_to_install_incompatible_error">%1$s could not be installed because it is not compatible with %2$s %3$s.</string>
     <!-- Text shown when attempting to install a blocklisted add-on. %1$s is the add-on name. -->
-    <string name="mozac_feature_addons_blocklisted">%1$s could not be installed because it has a high risk of causing stability or security problems</string>
+    <string name="mozac_feature_addons_blocklisted">%1$s could not be installed because it has a high risk of causing stability or security problems.</string>
```
@Amejia481 
This is a string change and it requires a new ID
https://mozilla-l10n.github.io/documentation/localization/making_string_changes.html
